src/dex/aave-v3-stata/aave-v3-stata.ts

```diff
@@ -0,0 +1,507 @@
+      return null;
+    }
+
+    const isSrcStata = src === TokenType.STATA_TOKEN;
+
+    const [stataToken, otherAddressLower] = isSrcStata
+      ? [
+          getTokenFromAddress(this.network, srcToken.address),
+          destToken.address.toLowerCase(),
+        ]
+      : [
+          getTokenFromAddress(this.network, destToken.address),
+          srcToken.address.toLowerCase(),
+        ];
+
+    // the token itself can only swap from/to underlying and aToken, so
+    // - at least one must be stata
+    // - maximum one can be stata
+    // - second one must be underlying or aUnderlying
+    // on the buy side (mint, withdraw) we only support the underlying<->stata conversion, not the aUnderlying
+    if (side === SwapSide.SELL) {
+      if (
+        otherAddressLower !== stataToken.underlying.toLowerCase() &&
+        otherAddressLower !== stataToken.underlyingAToken.toLowerCase()
+      ) {
+        return null;
+      }
+    } else {
+      if (otherAddressLower !== stataToken.underlying.toLowerCase()) {
+        return null;
+      }
+    }
+
+    const stataAddressLower = stataToken.address.toLowerCase();
+
+    if (
+      !this.state[stataAddressLower]?.blockNumber ||
+      blockNumber > this.state[stataAddressLower].blockNumber
+    ) {
+      const cached = await this.dexHelper.cache.get(
+        this.dexKey,
+        this.network,
+        `state_${stataAddressLower}`,
+      );
+      if (cached) {
+        this.state[stataAddressLower] = Utils.Parse(cached);
+      } else {
+        const results = await this.dexHelper.multiWrapper.tryAggregate<bigint>(
+          true,
+          [
+            {
+              target: stataToken.address,
+              callData: AaveV3Stata.stata.encodeFunctionData('rate'),
+              decodeFunction: uint256ToBigInt,
+            },
+          ],
+          blockNumber,
+        );
+        this.state[stataAddressLower] = {
+          blockNumber,
+          rate: results[0].returnData,
+        };
+        this.dexHelper.cache.setex(
+          this.dexKey,
+          this.network,
+          `state_${stataAddressLower}`,
+          60,
+          Utils.Serialize(this.state[stataAddressLower]),
+        );
+      }
+    }
+
+    return [
+      {
+        prices: amounts.map(amount => {
+          const rate = this.state[stataAddressLower].rate;
+          if (side === SwapSide.SELL) {
+            if (isSrcStata) {
+              return this.previewRedeem(amount, rate);
+            } else {
+              return this.previewDeposit(amount, rate);
+            }
+          } else {
+            if (isSrcStata) {
+              return this.previewWithdraw(amount, rate);
+            } else {
+              return this.previewMint(amount, rate);
+            }
+          }
+        }),
+        unit: getBigIntPow(
+          (side === SwapSide.SELL ? destToken : srcToken).decimals,
+        ),
+        gasCost: 400_000, // 250_000 from underlying, far less from aToken
```

Comment:
Redundantly high gasCost could deprioritize pricing against other dexes, better to leave todo and update after tests

Reply:
updated